### PR TITLE
Entry deletion for the simple workflow

### PR DESCRIPTION
**Pull request by Benaiah**
**- Summary**

This PR adds a "Delete" button to the entry page in the simple workflow which deletes the current entry from the backend. This does not close https://github.com/netlify/netlify-cms/issues/291 completely, but it is most of the way there (the remaining work to close that ticket is to implement deletion in the editorial workflow, which is in progress in the `delete-entries-in-editorial-workflow` branch.

**- Test plan**

Existing tests pass, relevant spec tests were updated, and the PR was manually tested with editorial workflow both enabled and disabled.

**- Description for the changelog**

Entry deletion for the simple workflow

**- A picture of a cute animal (not mandatory but encouraged)**

image

### CODE REVIEW

**Comment by erquhart** [comment]
> the remaining work to close that ticket is to implement deletion in the editorial workflow

@Benaiah does this PR regress the current editorial workflow deletion functionality? Or is the other branch improving that functionality?

File: src/containers/editorialWorkflow/EntryPageHOC.js [hunk 1]
```diff
@@ -15,8 +15,9 @@ export default function EntryPageHOC(EntryPage) {
   function mapStateToProps(state, ownProps) {
     const { collections } = state;
     const isEditorialWorkflow = (state.config.get('publish_mode') === EDITORIAL_WORKFLOW);
-    const returnObj = { isEditorialWorkflow };
+    const returnObj = { isEditorialWorkflow, showDelete: true };
```
Comment (erquhart) [inline src/containers/editorialWorkflow/EntryPageHOC.js]: `showDelete` should also be false if the entry is new.

File: src/containers/EntryPage.js [hunk 2]
```diff
@@ -79,16 +82,29 @@ class EntryPage extends React.Component {
...
   handleCloseEntry = () => {
-    this.props.closeEntry();
+    return this.props.closeEntry();
   };
 
   handlePersistEntry = () => {
     const { persistEntry, collection } = this.props;
     setTimeout(() => {
-      persistEntry(collection);  
+      persistEntry(collection).then(() => this.handleCloseEntry());
```
Comment (erquhart) [inline src/containers/EntryPage.js]: 👍

File: src/backends/github/implementation.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import semaphore from "semaphore";
+import { resolvePromiseProperties } from "../../lib/promiseHelper";
```
Comment (erquhart) [inline src/backends/github/implementation.js]: Unused import.

File: src/backends/github/API.js [hunk 4]
```diff
@@ -258,6 +263,23 @@ export default class API {
...
   }
 
+  deleteFile(path, message, options={}) {
+    const branch = options.branch || this.branch;
+    // We need to request the file first to get the SHA
+    return this.request(`${ this.repoURL }/contents/${ path }`)
+    .then(({ sha }) => this.request(
+      `${ this.repoURL }/contents/${ path }`,
+      {
+        method: "DELETE",
```
Comment (erquhart) [inline src/backends/github/API.js]: Go with single quotes for this and other quotes in the PR (except React component attributes).

File: src/backends/github/API.js [hunk 5]
```diff
@@ -258,6 +263,23 @@ export default class API {
     });
   }
 
+  deleteFile(path, message, options={}) {
+    const branch = options.branch || this.branch;
+    // We need to request the file first to get the SHA
+    return this.request(`${ this.repoURL }/contents/${ path }`)
```
Comment (erquhart) [inline src/backends/github/API.js]: Nitpick: we can store the assembled url above the return and reuse.

File: src/backends/github/API.js [hunk 6]
```diff
@@ -343,10 +365,18 @@ export default class API {
...
-    let prNumber; 
     return this.retrieveMetadata(contentKey)
     .then(metadata => this.closePR(metadata.pr, metadata.objects))
-    .then(() => this.deleteBranch(`cms/${ contentKey }`));
+    .then(() => this.deleteBranch(`cms/${ contentKey }`))
+    // If the PR doesn't exist, then this has already been deleted -
+    // deletion should be idempotent, so we can consider this a
+    // success.
+    .catch((err) => {
+      if (err.message === "Reference does not exist") {
```
Comment (erquhart) [inline src/backends/github/API.js]: Nothing less ephemeral to check against, such as an error code?

Comment (Benaiah) [inline src/backends/github/API.js]: The error code here is a 422, and there's no other metadata available there other than the error message. [422 is a very general error code in the GitHub API](https://developer.github.com/v3/#client-errors), so resolving this promise for all 422s will other errors, so I checked the more specific error message.

https://developer.github.com/v3/#client-errors says clients should differentiate 422 errors based on the body of the response (the `APIError` wrapper we use isn't including this information, which should probably be changed in another PR). I was hoping there would be more metadata in the request, but when inspecting a relevant 422 response manually, only the `message` and `documentation_url` were present, leaving us with just the `message` and status to switch between. Given GitHub's API documentation, it seems that the `message` is intended to be used to differentiate errors, so I used that.

If you'd rather catch all 422 errors, I can change this pretty trivially (just need to change the condition to `err.status === 422`).

File: src/containers/editorialWorkflow/EntryPageHOC.js (hunk 1 above)
Comment (Benaiah) [inline src/containers/editorialWorkflow/EntryPageHOC.js]: Currently it deletes your unsaved changes - that seemed like the most obvious approach to me. I'll switch this out.

File: src/backends/github/implementation.js (hunk 3 above)
Comment (Benaiah) [inline src/backends/github/implementation.js]: Thanks for the catch!

File: src/backends/github/API.js (hunk 4 above)
Comment (Benaiah) [inline src/backends/github/API.js]: 👍

File: src/backends/github/API.js (hunk 6 above)
Comment (erquhart) [inline src/backends/github/API.js]: Okay, this is fine as is. Thanks for the explanation.

File: src/containers/editorialWorkflow/EntryPageHOC.js (hunk 1 above)
Comment (erquhart) [inline src/containers/editorialWorkflow/EntryPageHOC.js]: Is this change still on the way?

**Comment by erquhart** [comment]
@carlflor would you be up for reviewing this PR? You mentioned contributing, and code review is one area that needs much more community activity, and we prefer to have two maintainers/collaborators okay a PR before merging. Let me know!

**Comment by carlflor** [comment]
@erquhart okay! cool! I can review this :)

**Comment by Benaiah** [comment]
@carlflor thanks for the review!

**Comment by mbergeronupgrade** [comment]
@Benaiah I think in editorial mode, the delete entry button does not respect the `delete: false` option from `config.yml`

**Pull request merged**